layout/layout_test.go

```diff
@@ -0,0 +1,486 @@
+					h.AssertEq(t, len(index.Manifests), 1)
+					h.AssertEq(t, "my-additional-tag", index.Manifests[0].Annotations["org.opencontainers.image.ref.name"])
+				})
+
+				it("failed on saved when more than one name is provided", func() {
+					image, err := layout.NewImage(imagePath, layout.FromBaseImage(testImage))
+					h.AssertNil(t, err)
+
+					// save on disk in OCI
+					err = image.Save("name1", "name2")
+					h.AssertError(t, err, "multiple additional names [name1 name2] are not allow when OCI layout is used")
+				})
+			})
+
+			when("no additional names are provided", func() {
+				it("creates an image with all the layers from the underlying image", func() {
+					image, err := layout.NewImage(imagePath, layout.FromBaseImage(testImage))
+					h.AssertNil(t, err)
+
+					// save on disk in OCI
+					err = image.Save()
+					h.AssertNil(t, err)
+
+					//  expected blobs: manifest, config, layer
+					h.AssertBlobsLen(t, imagePath, 3)
+
+					// assert additional name
+					index := h.ReadIndexManifest(t, imagePath)
+					h.AssertEq(t, len(index.Manifests), 1)
+					h.AssertEq(t, 0, len(index.Manifests[0].Annotations))
+				})
+			})
+		})
+
+		when("#FromBaseImagePath", func() {
+			it.Before(func() {
+				imagePath = filepath.Join(tmpDir, "save-from-base-image-path")
+			})
+
+			when("full image was saved on disk in OCI layout format", func() {
+				when("a new layer was added", func() {
+					it("image is saved on disk with all the layers", func() {
+						image, err := layout.NewImage(imagePath, layout.FromBaseImagePath(fullBaseImagePath))
+						h.AssertNil(t, err)
+
+						// add a random layer
+						path, diffID, _ := h.RandomLayer(t, tmpDir)
+						err = image.AddLayerWithDiffID(path, diffID)
+						h.AssertNil(t, err)
+
+						// save on disk in OCI
+						err = image.Save()
+						h.AssertNil(t, err)
+
+						// expected blobs: manifest, config, base image layer, new random layer
+						h.AssertBlobsLen(t, imagePath, 4)
+					})
+				})
+			})
+
+			when("sparse image was saved on disk in OCI layout format", func() {
+				when("a new layer was added", func() {
+					it("image is saved on disk with the new layer only", func() {
+						image, err := layout.NewImage(imagePath, layout.FromBaseImagePath(sparseBaseImagePath))
+						h.AssertNil(t, err)
+
+						// add a random layer
+						path, diffID, _ := h.RandomLayer(t, tmpDir)
+						err = image.AddLayerWithDiffID(path, diffID)
+						h.AssertNil(t, err)
+
+						// save on disk in OCI
+						err = image.Save("v1.0")
+						h.AssertNil(t, err)
+
+						// expected blobs: manifest, config, new random layer
+						h.AssertBlobsLen(t, imagePath, 3)
+
+						// assert additional name
+						index := h.ReadIndexManifest(t, imagePath)
+						h.AssertEq(t, len(index.Manifests), 1)
+						h.AssertEq(t, "v1.0", index.Manifests[0].Annotations["org.opencontainers.image.ref.name"])
+					})
+				})
+			})
+		})
+
+		when("#FromPreviousImage", func() {
```

Comment:
What if the previous image is sparse?

Reply:
I think I missed that test case, we can't reuse the layer if it doesn't exist from the previous image (because it is sparse) so I think we should err but not sure.

Reply:
Well, it seems to be working, the layers are added into the Manifest but they are not present in the `blobs`